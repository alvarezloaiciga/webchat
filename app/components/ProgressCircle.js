// @flow

import React from 'react';
import './styles/ProgressCircle.scss';

export type ProgressBarProps = {
  percentage: number,
  size: number,
  progressColor: string,
  trackColor: string,
  progressWidth: number,
  trackWidth: number,
};

export type ProgressBarState = {
  currentPercentage: number,
  targetPercentage: number,
};

const circleTop = Math.PI / -2;
const circumference = 2 * Math.PI;

export class ProgressCircle extends React.Component<ProgressBarProps, ProgressBarState> {
  static defaultProps = {
    progressWidth: 8,
    trackWidth: 8,
    progressColor: '#fff',
    trackColor: 'rgba(128, 128, 128, 0.5)',
    size: 50,
  };

  canvas: HTMLCanvasElement;
  targetPercentage: number;
  currentPercentage: number = 0;
  origin: number;
  radius: number;
  animating: boolean = false;

  constructor(props: ProgressBarProps) {
    super(props);
    this.targetPercentage = props.percentage;
    this.origin = props.size / 2;
    this.radius = Math.max(
      (props.size - 2 * Math.max(props.progressWidth, props.trackWidth)) / 2,
      0,
    );
  }

  componentWillReceiveProps(newProps: ProgressBarProps) {
    this.targetPercentage = newProps.percentage;
    this.origin = newProps.size / 2;
    this.radius = Math.max(
      (newProps.size - 2 * Math.max(newProps.progressWidth, newProps.trackWidth)) / 2,
      0,
    );
    if (!this.animating) {
      this.animate(null, true);
    }
  }

  componentDidMount() {
    this.animate();
  }

  drawTrack = (ctx: CanvasRenderingContext2D) => {
    ctx.lineWidth = this.props.trackWidth;
    ctx.strokeStyle = this.props.trackColor;
    ctx.beginPath();
    ctx.arc(this.origin, this.origin, this.radius, 0, 2 * Math.PI, false);
    ctx.stroke();
  };

  drawProgress = (ctx: CanvasRenderingContext2D, toRadian: number) => {
    ctx.lineWidth = this.props.progressWidth;
    ctx.strokeStyle = this.props.progressColor;
    ctx.beginPath();
    ctx.arc(this.origin, this.origin, this.radius, circleTop, toRadian, false);
    ctx.stroke();
  };

  animate = (highResTimestamp: ?number, forceDraw: boolean = false) => {
    // Bail out if we're at the target percentage and we don't want to force drawing
    if (!forceDraw && this.currentPercentage === this.targetPercentage) {
      this.animating = false;
      return;
    }

    this.animating = true;

    const ctx = this.canvas && this.canvas.getContext('2d');

    // Bail out if ctx is null--ex. device can't support canvas (extremely unlikely)
    if (!ctx) {
      return;
    }

    // Bump to next percentage
    if (this.targetPercentage !== this.currentPercentage) {
      this.currentPercentage =
        this.targetPercentage > this.currentPercentage
          ? this.currentPercentage + 1
          : this.currentPercentage - 1;
    }

    const nextRadian = this.currentPercentage / 100 * circumference + circleTop;

    // Clear off the canvas
    ctx.clearRect(0, 0, 100, 100);

    // Draw track
    this.drawTrack(ctx);

    // Draw progress
    this.drawProgress(ctx, nextRadian);

    // Recursively animate until end
    requestAnimationFrame(this.animate);
  };

  render() {
    return (
      <canvas
        ref={n => {
          if (n) this.canvas = n;
        }}
        width={this.props.size}
        height={this.props.size}
      />
    );
  }

  /*return (
    <div className="ProgressBar"
      style={{
        height: props.width ? `${props.width}px` : '10px',
      }}
    >
      <div className="track"
        style={{
          backgroundColor: props.trackColor,
        }}
      />
      <div className="progressBar"
        style={{
          width: `${Math.max(0, Math.min(props.percentage, 100))}%`,
          backgroundColor: props.progressColor,
        }}
      />
    </div>
  );*/
}

export default ProgressCircle;
