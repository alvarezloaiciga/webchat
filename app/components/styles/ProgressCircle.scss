.ProgressBar {
  width: 100%;
  display: block;
  padding: 0;
  position: relative;

  .track,
  .progressBar {
    display: block;
    border-radius: 9999px;
    height: 100%;
  }

  .track {
    width: 100%;
    background-color: rgba(255, 255, 255, 0.5);
  }

  .progressBar {
    background-color: white;
    position: absolute;
    left: 0;
    top: 0;
    transition: width 3s;
  }
}
