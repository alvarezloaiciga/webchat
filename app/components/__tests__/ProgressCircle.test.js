import React from 'react';
import {ProgressCircle} from '../ProgressCircle';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';

describe('ProgressCircle component', () => {
  let wrapper: ShallowWrapper;
  const testProps = {
    percentage: 50,
    size: 60,
    progressColor: '#fff',
    trackColor: '#000',
    progressWidth: 3,
    trackWidth: 4,
  };

  describe('rendering', () => {
    it('renders', () => {
      wrapper = shallow(<ProgressCircle {...testProps} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
