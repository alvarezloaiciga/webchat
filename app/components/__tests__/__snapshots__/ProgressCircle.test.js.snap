// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressCircle component rendering renders 1`] = `
<canvas
  height={60}
  width={60}
/>
`;
